servlet-engine/src/main/java/io/micronaut/servlet/engine/body/ExtendedInputStream.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.servlet.engine.body;
+
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.http.body.ByteBody;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Extended InputStream API for better backpressure/cancellation handling.
+ *
+ * @author Jonas Konrad
+ * @since 4.9.0
+ */
+abstract class ExtendedInputStream extends InputStream {
```

Comment:
make sealed

Reply:
The implementations in StreamPair are private so I can't add permits clauses